common/utils/include/claragenomics/utils/cudasort.cuh

```diff
@@ -0,0 +1,216 @@
+    // A limitation of CUB's implementation is that it moves values to their new location in every step. If values are larger than 4 bytes this
+    // begins to dominate. That's why it's better to use a 32-bit index and move the values to the final location in the end.
+
+    // CUB accepts the number of elements as int, check if array is longer than that
+    // (this limitation can be avoided by some CUB trickery, but no need to do it here)
+    if (values.size() > std::numeric_limits<int>::max())
+    {
+        throw std::length_error("cudasort: array too long to be sorted");
+    }
+    using move_to_index_t = std::uint32_t;
+
+    const auto number_of_elements = values.size();
+
+    thrust::device_vector<move_to_index_t> move_to_index(number_of_elements);
+    // Fill array with values 0..number_of_elements-1
+    thrust::sequence(thrust::device,
+                     std::begin(move_to_index),
+                     std::end(move_to_index));
+
+    // *** sort by less significant key first ***
+    thrust::device_vector<LessSignificantKeyT> less_significant_key_sorted(number_of_elements);
+    thrust::device_vector<move_to_index_t> move_to_index_sorted(number_of_elements);
+
+    thrust::device_vector<std::uint8_t> temp_storage_vect;
+
+    details::perform_radix_sort(temp_storage_vect,
+                                less_significant_keys.data().get(),
+                                less_significant_key_sorted.data().get(),
+                                move_to_index.data().get(),
+                                move_to_index_sorted.data().get(),
+                                number_of_elements,
+                                0,
+                                int_floor_log2(max_value_of_less_significant_key) + 1); // function asks for the index of past-the-last most significant bit
+
+    // swap sorted and unsorted arrays
+    thrust::swap(less_significant_keys, less_significant_key_sorted);
+    thrust::swap(move_to_index, move_to_index_sorted);
+
+    // deallocate helper array
+    // TODO: This array can probably be reused, but waiting for more general reallocation-avoidance strategy before optimizing this
+    less_significant_key_sorted.clear();
+    less_significant_key_sorted.shrink_to_fit();
+
+    // *** move more significant keys to their position after less significant keys sort ***
+    thrust::device_vector<MoreSignificantKeyT> more_significant_keys_after_sort(number_of_elements);
+
+    thrust::gather(thrust::device,
+                   std::begin(move_to_index),
+                   std::end(move_to_index),
+                   std::begin(more_significant_keys),
+                   std::begin(more_significant_keys_after_sort));
+
+    thrust::swap(more_significant_keys, more_significant_keys_after_sort);
+
+    // *** sort by more significant key ***
+
+    thrust::device_vector<std::uint8_t> temp_storage_vect_2;
+
+    details::perform_radix_sort(temp_storage_vect_2,
```

Comment:
Can't you reuse the temp storage from the first radix sort?

Reply:
Yes I can, in fact `perform_radix_sort` is implemented so that it reuse it (or resize if needed). Thanks for spotting

Reply:
done